core/src/main/scala/kafka/network/RequestChannel.scala

```diff
@@ -94,19 +104,64 @@ object RequestChannel extends Logging {
     @volatile var recordNetworkThreadTimeCallback: Option[Long => Unit] = None
 
     val session = Session(context.principal, context.clientAddress)
+
     private val bodyAndSize: RequestAndSize = context.parseRequest(buffer)
 
     def header: RequestHeader = context.header
     def sizeOfBodyInBytes: Int = bodyAndSize.size
 
-    //most request types are parsed entirely into objects at this point. for those we can release the underlying buffer.
-    //some (like produce, or any time the schema contains fields of types BYTES or NULLABLE_BYTES) retain a reference
-    //to the buffer. for those requests we cannot release the buffer early, but only when request processing is done.
+    // most request types are parsed entirely into objects at this point. for those we can release the underlying buffer.
+    // some (like produce, or any time the schema contains fields of types BYTES or NULLABLE_BYTES) retain a reference
+    // to the buffer. for those requests we cannot release the buffer early, but only when request processing is done.
     if (!header.apiKey.requiresDelayedAllocation) {
       releaseBuffer()
     }
 
-    def requestDesc(details: Boolean): String = s"$header -- ${loggableRequest.toString(details)}"
+    def isForwarded(): Boolean = envelopeContext.isDefined
+
+    def buildResponse(abstractResponse: AbstractResponse,
+                      error: Errors): Send = {
```

Comment:
I think this logic still conflates the envelope error and the inner response error. We might catch an exception raised from `validateForwardRequest` or from the request handler in `KafkaApis.handle`. Both paths flow through `KafkaApis.handleError`, so we do not have a way to distinguish the two cases. This means that an uncaught error from the underlying request will get sent back to the forwarded broker as an error in the envelope, which will cause us to translate it to UNKNOWN_SERVER_ERROR.

I think we should handle envelope errors explicitly through a separate method. We can define a separate method here such as `buildFailedEnvelope` which can be used inside `validateForwardRequest`. Then inside `buildResponse` here, we can always return `Errors.NONE` as the envelope error.